src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+understand correctly, the proxy handles all requests that are to be
+forwarded to individual compute nodes (such as requesting files, or
+resources on web servers belonging to a specific project).  HAproxy doesn't
+know anything about the compute nodes themseves--it just sees URLs that look
+like they belong to a project (they begin with a project UUID) and forwards
+those requests to the Hub's proxy, which in turn checks that the requester
+is authenticated and has permissions to access that project's resources.
+The proxy then forwards the request to the appropriate port on that
+project's compute node::
+
+    Client <--> HAProxy <--> Hub Proxy <--> Compute
+
+Compute
+^^^^^^^
+
+Compute servers are where the real work gets done in SMC projects.  Every
+project is associated with a specific compute server where all their data
+is stored (by way of storage servers mounted on the compute node) and where
+all process and computation tasks done by the project are performed.  This
+includes running Sage.  The Compute servers are Linux VMs with varying
+degrees of hardware capacity, depending on how much you're willing to pay.
+In most cases the servers are shared between projects (you don't have admin
+on the servers) though in principle one could pay for one's own compute
+server as well.
+
+Otherwise, one can do quite a bit of different things on their compute node,
+including log in to the shell directly (you log in as a user named after
+your project's UUID).  This can be done either through the web terminal in
+SMC, or one can SSH in directly.
+
+Each compute node also runs a simple socket server
+``smc-hub/compute-server.coffee`` that is used by the Hub to communicate with
+the compute node (using simple JSON messages). For example, one can make status
+inquiries on the node, or send commands to run a command in a project.
+
+Additionally, each *project* runs a couple per-project daemons when the
+project is created and running.  These include the the console server
+(``smc-project/console_server.coffee``) which provides the backend for the
+web terminal, and the "local hub" (``smc-project/local_hub.coffee``).  We
+haven't yet detailed everything that the "local hub" does, but whereas the
+"compute-server" manages the entire compute node, the "local hub" runs
+per-project (as that project's user) and helps coordinate connections
+between software running in the project and the "global hub" (i.e. the Hub,
+through which the client is communicating).
+
+As we understand it, whereas the compute server is used to issue commands to
+the compute node on behalf of a project, such as starting the project's
+local hub, the local hub then takes care of the rest.  This all makes sense,
+but gets rather complicated as starting the actual local hub is buried under
+a pile of Python scripts.  Some of this may still be legacy from earlier
+versions of SMC that were written primarily in Python, and is need of
+cleaning up.
```

Comment:
There's an unfinished rewrite on top of kubernetes... that is even more complicated, but is the right way to go.  That rewrite, which finished, isn't going to be open source, since it'll only be needed by very large sites....